### PR TITLE
`run` macro for Scala 3

### PR DESCRIPTION
**Pull request by lrytz**
The `RunningSpec` test has been simplified a bit because not all combinators are available yet

### CODE REVIEW

File: parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala [hunk 1]
```diff
@@ -334,6 +335,72 @@ object ParserMacros {
...
+
+    case class RuleFromParser(parser: Term, rule: Symbol, targs: List[TypeTree], argss: List[List[Term]]):
+      def ruleCall[P](localParser: Expr[P]): Expr[RuleN[L]] =
+        val r = Select(localParser.asTerm, rule)
+        argss.foldLeft(if (targs.isEmpty) r else TypeApply(r, targs))((t, args) => Apply(t, args)).asExprOf[RuleN[L]]
+
+    object RuleFromParser:
+      def dissect(t: Term, targs: List[TypeTree], argss: List[List[Term]]): (Term, List[TypeTree], List[List[Term]]) =
+        t match
+          case Inlined(_, _, t)    => dissect(t, targs, argss)
```
Comment (jrudolph) [inline parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala:368]: I used `t.underlyingArgument` at other places to get rid of `Inlined` wrapping.

File: parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala [hunk 2]
```diff
@@ -334,6 +335,72 @@ object ParserMacros {
   import scala.quoted._
   import scala.compiletime._
 
+  // TODO: the `R` type parameter is a workaround for https://github.com/lampepfl/dotty/issues/13376
+  // Discussion at https://github.com/sirthias/parboiled2/pull/274#issuecomment-904926294
+  def runImpl[L <: HList: Type, R: Type]()(ruleExpr: Expr[RuleN[L]], schemeExpr: Expr[Parser.DeliveryScheme[L]])(using
+      Quotes
+  ): Expr[R] =
+    import quotes.reflect.*
```
Comment (jrudolph) [inline parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala:343]: We haven't used indentation-based syntax so far. I would prefer to keep braces for now (my IntelliJ gets completely messed up without), though it's not a strong opinion.

File: parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala [hunk 3]
```diff
@@ -334,6 +335,72 @@ object ParserMacros {
...
+            // This is why `asInstanceOf[Parser]` is needed below
+            val parserExpr = rule.parser.asExprOf[pT]
+            '{
+              val p: pT = $parserExpr
+              p.asInstanceOf[Parser].__run[L](${ rule.ruleCall('p) })($schemeExpr).asInstanceOf[R]
+            }
+      case Inlined(_, _, Inlined(Some(ruleMacro), List(ValDef(_, _, Some(parserThis))), rule))
+          if rule.tpe <:< TypeRepr.of[RuleX] && isRuleMacro(ruleMacro.symbol) =>
+        // The `Inlined` tree for the `rule` macro has a binding for the parser instance.
+        // TODO: we re-use the rhs of that binding (parserThis), I didn't manage to create the right This() tree.
```
Comment (jrudolph) [inline parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala:395]: Interesting, I also cannot think of something better right now.

File: parboiled-core/src/test/scala/org/parboiled2/RunningSpec.scala [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+import scala.util.{Success, Try}
+import utest._
+
+object RunningSpec extends TestSuite {
+
+  class TestParser(val input: ParserInput) extends Parser {
+    //def A                 = rule('a' ~ B ~ EOI)
+    //def B                 = rule(oneOrMore('b'))
+    //def C(n: Int)(m: Int) = rule((n - m).times('c'))
```
Comment (jrudolph) [inline parboiled-core/src/test/scala/org/parboiled2/RunningSpec.scala:27]: In #287, I have ported lots more combinator, only `times` is missing for this test, so I'm trying to add that as well now.

**Review by jrudolph — COMMENTED** [review]
Some comments below.

File: parboiled-core/src/main/scala-2/org/parboiled2/ParserMacros.scala [hunk 5]
```diff
@@ -56,11 +56,10 @@ object ParserMacros {
     val runCall = c.prefix.tree match {
       case q"parboiled2.this.Rule.Runnable[$l]($ruleExpr)" =>
         ruleExpr match {
-          case q"$p.$r" if p.tpe <:< typeOf[Parser]        => q"val p = $p; p.__run[$l](p.$r)($scheme)"
-          case q"$p.$r($args)" if p.tpe <:< typeOf[Parser] => q"val p = $p; p.__run[$l](p.$r($args))($scheme)"
-          case q"$p.$r[$t]" if p.tpe <:< typeOf[Parser]    => q"val p = $p; p.__run[$l](p.$r[$t])($scheme)"
-          case q"$p.$r[$t]" if p.tpe <:< typeOf[RuleX]     => q"__run[$l]($ruleExpr)($scheme)"
-          case x                                           => c.abort(x.pos, "Illegal `.run()` call base: " + x)
+          case q"$p.$r[..$ts](...$argss)" if p.tpe <:< typeOf[Parser] =>
```
Comment (jrudolph) [inline parboiled-core/src/main/scala-2/org/parboiled2/ParserMacros.scala:59]: :+1: nice!

File: parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala (hunk 2 above)
Comment (lrytz) [reply parboiled-core/src/main/scala-3/org/parboiled2/ParserMacros.scala:343]: OK, sure. I'll add the braces.

**Comment by jrudolph** [comment]
#287 is merged, can you do another rebase?

**Comment by lrytz** [comment]
This is done (at least for now)

**Review by jrudolph — APPROVED** [review]
Great, thanks, @lrytz!

**Pull request merged**